pilot/pkg/model/push_context.go

```diff
@@ -1432,21 +1438,30 @@ func SortServicesByCreationTime(services []*Service) []*Service {
 // Caches list of service accounts in the registry
 func (ps *PushContext) initServiceAccounts(env *Environment, services []*Service) {
 	for _, svc := range services {
-		if ps.ServiceAccounts[svc.Hostname] == nil {
-			ps.ServiceAccounts[svc.Hostname] = map[int][]string{}
-		}
 		for _, port := range svc.Ports {
 			if port.Protocol == protocol.UDP {
 				continue
 			}
-			s, f := env.EndpointIndex.ShardsForService(string(svc.Hostname), svc.Attributes.Namespace)
-			if !f {
-				continue
-			}
-			s.RLock()
-			sa := spiffe.ExpandWithTrustDomains(s.ServiceAccounts, ps.Mesh.TrustDomainAliases).SortedList()
-			s.RUnlock()
-			ps.ServiceAccounts[svc.Hostname][port.Port] = sa
+			var accounts sets.Set
+			func() {
+				// First get endpoint level service accounts
+				shard, f := env.EndpointIndex.ShardsForService(string(svc.Hostname), svc.Attributes.Namespace)
+				if f {
+					shard.RLock()
+					defer shard.RUnlock()
+					accounts = shard.ServiceAccounts
+				}
+				if len(svc.ServiceAccounts) > 0 {
```

Comment:
lock on service?

Reply:
I am not sure what you mean?

Reply:
Earlier we used to have a lock on service before reading service accounts like 
```

	s.RLock()
	sa := spiffe.ExpandWithTrustDomains(s.ServiceAccounts, ps.Mesh.TrustDomainAliases).SortedList()
	s.RUnlock()
```

Reply:
thats from the `shard` (renamed from `s` to avoid confusion with service -- you are commenting on the model.Service. The shard is still locked

Reply:
I see. Makes sense